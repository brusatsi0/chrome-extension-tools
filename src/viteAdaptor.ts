import { EmittedFile, PluginContext } from 'rollup'
import { interpret } from 'xstate'
import { narrowEvent } from './files_helpers'
import { isFunction, isString, isUndefined } from './helpers'
import { RPCEPlugin } from './types'
import {
  configHook,
  getEmittedFileId,
  serverHook,
  viteAdaptorMachine,
  viteAdaptorModel,
} from './viteAdaptor.machine'

const service = interpret(viteAdaptorMachine)

service.start()

// service.subscribe((state) => {
//   console.log('ðŸš€ ~ state', state)
// })

const sendEmitFile: PluginContext['emitFile'] = (
  file: EmittedFile,
) => {
  service.send(viteAdaptorModel.events.EMIT_FILE(file))
  return getEmittedFileId(file)
}

export const fileWriteComplete = () =>
  new Promise<void>((resolve, reject) => {
    const sub = service.subscribe({
      next(state) {
        if (state.matches({ serve: { listening: 'ready' } })) {
          sub.unsubscribe()
          resolve()
        } else if (state.matches({ serve: 'error' })) {
          const { error, id } = narrowEvent(state.event, 'ERROR')
          error.id = id
          sub.unsubscribe()
          reject(error)
        }
      },
      error(err) {
        reject(err)
      },
      complete() {
        reject(
          new Error(`The service "${service.id}" has stopped`),
        )
      },
    })
  })

export const shimPluginContext = (
  pluginContext: PluginContext,
): PluginContext => {
  if (!service.initialized) return pluginContext

  const proxy = new Proxy(pluginContext, {
    get(target, key, receiver) {
      switch (key) {
        case 'emitFile':
          if (service.initialized) {
            return sendEmitFile
          }

        // eslint-disable-next-line no-fallthrough
        default:
          return Reflect.get(target, key, receiver)
      }
    },
  })

  return proxy
}

export const useViteAdaptor = (plugin: RPCEPlugin) => {
  service.send(viteAdaptorModel.events.ADD_PLUGIN(plugin))

  return new Proxy(plugin, {
    ownKeys(target) {
      const keys = new Set(Reflect.ownKeys(target))
      keys.add(configHook)
      keys.add(serverHook)
      return Array.from(keys)
    },
    get(target, prop, receiver) {
      const value = Reflect.get(target, prop, receiver)

      if (isString(prop) && isFunction(value))
        return function (this: PluginContext, ...args: any[]) {
          if (!service.initialized)
            return value.call(this, ...args)

          service.send(
            viteAdaptorModel.events.HOOK_START(prop, args),
          )

          const shim = shimPluginContext(this)

          return value.call(shim, ...args)
        }
      else if (
        (configHook === prop || serverHook === prop) &&
        isUndefined(value)
      )
        return function (this: PluginContext, ...args: any[]) {
          if (service.initialized)
            service.send(
              viteAdaptorModel.events.HOOK_START(prop, args),
            )
        }

      return value
    },
  })
}
