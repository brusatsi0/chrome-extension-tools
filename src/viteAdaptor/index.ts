import fs from 'fs-extra'
import path from 'path'
import { EmittedChunk } from 'rollup'
import { EmittedAsset } from 'rollup'
import { EmittedFile, PluginContext } from 'rollup'
import { from } from 'rxjs'
import { ViteDevServer } from 'vite'
import {
  ActorRef,
  createMachine,
  interpret,
  spawn,
} from 'xstate'
import { assign, pure, send } from 'xstate/lib/actions'
import { createModel } from 'xstate/lib/model'
import { ModelEventsFrom } from 'xstate/lib/model.types'
import { isString, isUndefined } from '../helpers'
import { bundleChunk } from './bundleChunk'

export const VITE_SERVER_URL = '__VITE_SERVER_URL__'
export const viteServerUrlRegExp = new RegExp(
  VITE_SERVER_URL,
  'g',
)

interface Context {
  server?: ViteDevServer
  pluginContext?: PluginContext
  lastError?: Error
  files: (EmittedFile & {
    id: string
    done?: true
    error?: string
  })[]
  writers: Record<
    string,
    ActorRef<ModelEventsFrom<typeof model>>
  >
  bundlers: Record<
    string,
    ActorRef<ModelEventsFrom<typeof model>>
  >
}

const getEmittedFileId = (file: EmittedFile): string =>
  file.type === 'asset'
    ? file.fileName ?? file.name ?? 'fake asset id'
    : file.id

const context: Context = {
  files: [],
  writers: {},
  bundlers: {},
}

const model = createModel(context, {
  events: {
    WRITE_ASSET: (file: EmittedAsset & { id: string }) => ({
      file,
    }),
    WRITE_CHUNK: (file: EmittedChunk) => ({ file }),
    EMIT_FILE: (file: EmittedFile) => ({
      file: {
        id: getEmittedFileId(file),
        ...file,
      },
    }),
    EMIT_DONE: (id: string) => ({ id }),
    HOOK_START: (hookName: string) => ({ hookName }),
    ERROR: (error: any, id?: string) => ({ id, error }),
    SERVER_CONFIGURE: (server: ViteDevServer) => ({ server }),
    SERVER_LISTENING: () => ({}),
  },
})

const machine = createMachine<typeof model>(
  {
    id: 'pluginContextShim',
    context: model.initialContext,
    initial: 'start',
    states: {
      start: {
        on: {
          HOOK_START: 'build',
          SERVER_CONFIGURE: {
            target: 'serve',
            actions: model.assign({
              server: (context, { server }) => server,
            }),
          },
        },
      },
      build: {
        type: 'final',
      },
      serve: {
        initial: 'starting',
        states: {
          error: { id: 'error' },
          starting: {
            invoke: {
              src: 'waitForServer',
            },
            on: {
              SERVER_LISTENING: 'listening',
              ERROR: {
                target: '#error',
                actions: model.assign({
                  lastError: (context, { error }) => error,
                }),
              },
              EMIT_FILE: {
                actions: model.assign({
                  files: ({ files }, { file }) => [
                    ...files,
                    file,
                  ],
                }),
              },
            },
          },
          listening: {
            entry: pure(({ files }) =>
              files.map((file) =>
                send(
                  file.type === 'asset'
                    ? model.events.WRITE_ASSET(file)
                    : model.events.WRITE_CHUNK(file),
                ),
              ),
            ),
            initial: 'writing',
            states: {
              writing: {},
              check: {
                always: [
                  {
                    cond: ({ files }) => {
                      return files.every(({ done }) => done)
                    },
                    target: 'ready',
                  },
                  'writing',
                ],
              },
              ready: {},
            },
            on: {
              EMIT_FILE: {
                target: '.writing',
                actions: [
                  'addFile',
                  send((context, { file }) =>
                    file.type === 'asset'
                      ? model.events.WRITE_ASSET(file)
                      : model.events.WRITE_CHUNK(file),
                  ),
                ],
              },
              WRITE_ASSET: { actions: 'writeAsset' },
              WRITE_CHUNK: { actions: 'writeChunk' },
              EMIT_DONE: {
                target: '.check',
                actions: model.assign({
                  files: ({ files }, { id }) => {
                    return files.map((file) =>
                      file.id === id
                        ? { ...file, done: true }
                        : file,
                    )
                  },
                }),
              },
              ERROR: {
                target: '#error',
                actions: model.assign({
                  files: ({ files }, { id, error }) =>
                    files.map((file) =>
                      file.id === id
                        ? { ...file, error, done: true }
                        : file,
                    ),
                  lastError: (context, { error }) => error,
                }),
              },
            },
          },
        },
      },
    },
  },
  {
    actions: {
      writeChunk: pure(({ server }, event) => {
        try {
          if (event.type !== 'WRITE_CHUNK')
            throw new Error(`Invalid event type: ${event.type}`)

          const { fileName, id } = event.file

          if (isUndefined(fileName))
            throw new Error(
              'EmittedChunk.fileName must be defined in vite serve',
            )

          if (isUndefined(server))
            throw new Error('vite server is undefined')

          return assign<Context, ModelEventsFrom<typeof model>>({
            bundlers: ({ bundlers }) => {
              const actorRef = spawn(
                from(
                  bundleChunk(event.file, server)
                    .then(() => model.events.EMIT_DONE(id))
                    .catch((error) =>
                      model.events.ERROR(id, error),
                    ),
                ),
                id,
              )
              return { ...bundlers, [id]: actorRef }
            },
          })
        } catch (error) {
          return send(
            model.events.ERROR(
              error,
              event.type === 'WRITE_CHUNK'
                ? event.file.id
                : event.type,
            ),
          )
        }
      }),
      writeAsset: pure(({ server }, event) => {
        try {
          if (event.type !== 'WRITE_ASSET')
            throw new Error(`Invalid event type: ${event.type}`)

          const { fileName, source, id } = event.file

          if (isUndefined(fileName))
            throw new Error(
              'EmittedAsset.fileName must be defined in vite serve',
            )
          if (isUndefined(source))
            throw new Error(
              'EmittedAsset.source must be defined in vite serve',
            )

          if (isUndefined(server))
            throw new Error('vite server is undefined')

          const filePath = path.join(
            server.config.build.outDir,
            fileName,
          )

          const {
            port,
            host = 'localhost',
            https,
          } = server.config.server

          if (isUndefined(port))
            throw new TypeError('vite server port is undefined')

          const fileSource = isString(source)
            ? source.replace(
                viteServerUrlRegExp,
                `${https ? 'https' : 'http'}://${host}:${port}`,
              )
            : source

          return assign<Context, ModelEventsFrom<typeof model>>({
            writers: ({ writers }) => {
              const actorRef = spawn(
                from(
                  fs
                    .outputFile(filePath, fileSource)
                    .then(() => model.events.EMIT_DONE(id))
                    .catch((error) =>
                      model.events.ERROR(id, error),
                    ),
                ),
                id,
              )
              return { ...writers, [id]: actorRef }
            },
          })
        } catch (error) {
          return send(
            model.events.ERROR(
              error,
              event.type === 'WRITE_ASSET'
                ? event.file.id
                : event.type,
            ),
          )
        }
      }),
    },
    services: {
      waitForServer: ({ server }) =>
        from(
          new Promise<ViteDevServer>((resolve, reject) => {
            if (!server?.httpServer)
              reject(
                new Error(
                  `vite httpServer is ${typeof server?.httpServer}`,
                ),
              )

            server?.httpServer?.once('listening', resolve)
          })
            .then(model.events.SERVER_LISTENING)
            .catch((error) =>
              model.events.ERROR(error, 'waitForServer'),
            ),
        ),
    },
  },
)

const service = interpret(machine)
service.start()

const sendEmitFile: PluginContext['emitFile'] = (
  file: EmittedFile,
) => {
  service.send(model.events.EMIT_FILE(file))
  return getEmittedFileId(file)
}

export const sendConfigureServer = (server: ViteDevServer) => {
  service.send(model.events.SERVER_CONFIGURE(server))
}

export const filesWritten = () =>
  new Promise<void>((resolve, reject) =>
    service.subscribe((state) => {
      if (state.matches({ serve: { listening: 'ready' } }))
        resolve()
      if (state.matches({ serve: 'error' }))
        reject(
          state.context.lastError ??
            `context.lastError is ${typeof state.context
              .lastError}`,
        )
    }),
  )

export const getViteServer = () => {
  const state = service.getSnapshot()
  return state.matches('serve')
    ? state.context.server
    : undefined
}

export const shimPluginContext = (
  pluginContext: PluginContext,
  hookName: string,
): PluginContext => {
  if (!service.initialized) return pluginContext

  service.send(model.events.HOOK_START(hookName))

  const proxy = new Proxy(pluginContext, {
    get(target, key, receiver) {
      switch (key) {
        case 'emitFile':
          if (service.initialized) {
            return sendEmitFile
          }

        // eslint-disable-next-line no-fallthrough
        default:
          return Reflect.get(target, key, receiver)
      }
    },
  })

  return proxy
}
