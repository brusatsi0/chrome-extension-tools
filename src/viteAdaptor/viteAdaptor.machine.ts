import fs from 'fs-extra'
import path from 'path'
import { EmittedChunk } from 'rollup'
import { EmittedAsset } from 'rollup'
import { EmittedFile, PluginContext } from 'rollup'
import { from } from 'rxjs'
import { ViteDevServer } from 'vite'
import { ActorRef, createMachine, spawn } from 'xstate'
import { assign, pure, send } from 'xstate/lib/actions'
import { createModel } from 'xstate/lib/model'
import { ModelEventsFrom } from 'xstate/lib/model.types'
import { isString, isUndefined } from '../helpers'
import { bundleChunk } from './bundleChunk'

export const VITE_SERVER_URL = '__VITE_SERVER_URL__'
export const viteServerUrlRegExp = new RegExp(
  VITE_SERVER_URL,
  'g',
)

interface Context {
  server?: ViteDevServer
  pluginContext?: PluginContext
  lastError?: Error
  files: (EmittedFile & {
    id: string
    done?: true
    error?: string
  })[]
  writers: Record<
    string,
    ActorRef<ModelEventsFrom<typeof viteAdaptorModel>>
  >
  bundlers: Record<
    string,
    ActorRef<ModelEventsFrom<typeof viteAdaptorModel>>
  >
}

export const getEmittedFileId = (file: EmittedFile): string =>
  file.type === 'asset'
    ? file.fileName ?? file.name ?? 'fake asset id'
    : file.id

const context: Context = {
  files: [],
  writers: {},
  bundlers: {},
}

export const viteAdaptorModel = createModel(context, {
  events: {
    WRITE_ASSET: (file: EmittedAsset & { id: string }) => ({
      file,
    }),
    WRITE_CHUNK: (file: EmittedChunk) => ({ file }),
    EMIT_FILE: (file: EmittedFile) => ({
      file: {
        id: getEmittedFileId(file),
        ...file,
      },
    }),
    EMIT_DONE: (id: string) => ({ id }),
    HOOK_START: (hookName: string) => ({ hookName }),
    ERROR: (error: any, id?: string) => ({ id, error }),
    SERVER_CONFIGURE: (server: ViteDevServer) => ({ server }),
    SERVER_LISTENING: () => ({}),
  },
})

export const viteAdaptorMachine = createMachine<
  typeof viteAdaptorModel
>(
  {
    id: 'viteAdaptor',
    context: viteAdaptorModel.initialContext,
    initial: 'start',
    states: {
      start: {
        on: {
          HOOK_START: 'build',
          SERVER_CONFIGURE: {
            target: 'serve',
            actions: viteAdaptorModel.assign({
              server: (context, { server }) => server,
            }),
          },
        },
      },
      build: {
        type: 'final',
      },
      serve: {
        initial: 'starting',
        states: {
          error: { id: 'error' },
          starting: {
            invoke: {
              src: 'waitForServer',
            },
            on: {
              SERVER_LISTENING: 'listening',
              ERROR: {
                target: '#error',
                actions: viteAdaptorModel.assign({
                  lastError: (context, { error }) => error,
                }),
              },
              EMIT_FILE: {
                actions: viteAdaptorModel.assign({
                  files: ({ files }, { file }) => [
                    ...files,
                    file,
                  ],
                }),
              },
            },
          },
          listening: {
            entry: pure(({ files }) =>
              files.map((file) =>
                send(
                  file.type === 'asset'
                    ? viteAdaptorModel.events.WRITE_ASSET(file)
                    : viteAdaptorModel.events.WRITE_CHUNK(file),
                ),
              ),
            ),
            initial: 'writing',
            states: {
              writing: {},
              check: {
                always: [
                  {
                    cond: ({ files }) => {
                      return files.every(({ done }) => done)
                    },
                    target: 'ready',
                  },
                  'writing',
                ],
              },
              ready: {},
            },
            on: {
              EMIT_FILE: {
                target: '.writing',
                actions: [
                  'addFile',
                  send((context, { file }) =>
                    file.type === 'asset'
                      ? viteAdaptorModel.events.WRITE_ASSET(file)
                      : viteAdaptorModel.events.WRITE_CHUNK(
                          file,
                        ),
                  ),
                ],
              },
              WRITE_ASSET: { actions: 'writeAsset' },
              WRITE_CHUNK: { actions: 'writeChunk' },
              EMIT_DONE: {
                target: '.check',
                actions: viteAdaptorModel.assign({
                  files: ({ files }, { id }) => {
                    return files.map((file) =>
                      file.id === id
                        ? { ...file, done: true }
                        : file,
                    )
                  },
                }),
              },
              ERROR: {
                target: '#error',
                actions: viteAdaptorModel.assign({
                  files: ({ files }, { id, error }) =>
                    files.map((file) =>
                      file.id === id
                        ? { ...file, error, done: true }
                        : file,
                    ),
                  lastError: (context, { error }) => error,
                }),
              },
            },
          },
        },
      },
    },
  },
  {
    actions: {
      writeChunk: pure(({ server }, event) => {
        try {
          if (event.type !== 'WRITE_CHUNK')
            throw new Error(`Invalid event type: ${event.type}`)

          const { fileName, id } = event.file

          if (isUndefined(fileName))
            throw new Error(
              'EmittedChunk.fileName must be defined in vite serve',
            )

          if (isUndefined(server))
            throw new Error('vite server is undefined')

          return assign<
            Context,
            ModelEventsFrom<typeof viteAdaptorModel>
          >({
            bundlers: ({ bundlers }) => {
              const actorRef = spawn(
                from(
                  bundleChunk(event.file, server)
                    .then(() =>
                      viteAdaptorModel.events.EMIT_DONE(id),
                    )
                    .catch((error) =>
                      viteAdaptorModel.events.ERROR(id, error),
                    ),
                ),
                id,
              )
              return { ...bundlers, [id]: actorRef }
            },
          })
        } catch (error) {
          return send(
            viteAdaptorModel.events.ERROR(
              error,
              event.type === 'WRITE_CHUNK'
                ? event.file.id
                : event.type,
            ),
          )
        }
      }),
      writeAsset: pure(({ server }, event) => {
        try {
          if (event.type !== 'WRITE_ASSET')
            throw new Error(`Invalid event type: ${event.type}`)

          const { fileName, source, id } = event.file

          if (isUndefined(fileName))
            throw new Error(
              'EmittedAsset.fileName must be defined in vite serve',
            )
          if (isUndefined(source))
            throw new Error(
              'EmittedAsset.source must be defined in vite serve',
            )

          if (isUndefined(server))
            throw new Error('vite server is undefined')

          const filePath = path.join(
            server.config.build.outDir,
            fileName,
          )

          const {
            port,
            host = 'localhost',
            https,
          } = server.config.server

          if (isUndefined(port))
            throw new TypeError('vite server port is undefined')

          const fileSource = isString(source)
            ? source.replace(
                viteServerUrlRegExp,
                `${https ? 'https' : 'http'}://${host}:${port}`,
              )
            : source

          return assign<
            Context,
            ModelEventsFrom<typeof viteAdaptorModel>
          >({
            writers: ({ writers }) => {
              const actorRef = spawn(
                from(
                  fs
                    .outputFile(filePath, fileSource)
                    .then(() =>
                      viteAdaptorModel.events.EMIT_DONE(id),
                    )
                    .catch((error) =>
                      viteAdaptorModel.events.ERROR(id, error),
                    ),
                ),
                id,
              )
              return { ...writers, [id]: actorRef }
            },
          })
        } catch (error) {
          return send(
            viteAdaptorModel.events.ERROR(
              error,
              event.type === 'WRITE_ASSET'
                ? event.file.id
                : event.type,
            ),
          )
        }
      }),
    },
    services: {
      waitForServer: ({ server }) =>
        from(
          new Promise<ViteDevServer>((resolve, reject) => {
            if (!server?.httpServer)
              reject(
                new Error(
                  `vite httpServer is ${typeof server?.httpServer}`,
                ),
              )

            server?.httpServer?.once('listening', resolve)
          })
            .then(viteAdaptorModel.events.SERVER_LISTENING)
            .catch((error) =>
              viteAdaptorModel.events.ERROR(
                error,
                'waitForServer',
              ),
            ),
        ),
    },
  },
)
