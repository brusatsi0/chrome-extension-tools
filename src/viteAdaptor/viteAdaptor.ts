import { EmittedFile, PluginContext } from 'rollup'
import { ViteDevServer } from 'vite'
import { interpret } from 'xstate'
import {
  viteAdaptorMachine,
  viteAdaptorModel,
  getEmittedFileId,
} from './viteAdaptor.machine'

const service = interpret(viteAdaptorMachine)

service.start()

const sendEmitFile: PluginContext['emitFile'] = (
  file: EmittedFile,
) => {
  service.send(viteAdaptorModel.events.EMIT_FILE(file))
  return getEmittedFileId(file)
}

export const sendConfigureServer = (server: ViteDevServer) => {
  service.send(viteAdaptorModel.events.SERVER_CONFIGURE(server))
}

export const filesWritten = () =>
  new Promise<void>((resolve, reject) =>
    service.subscribe((state) => {
      if (state.matches({ serve: { listening: 'ready' } }))
        resolve()
      if (state.matches({ serve: 'error' }))
        reject(
          state.context.lastError ??
            `context.lastError is ${typeof state.context
              .lastError}`,
        )
    }),
  )

export const getViteServer = () => {
  const state = service.getSnapshot()
  return state.matches('serve')
    ? state.context.server
    : undefined
}

export const shimPluginContext = (
  pluginContext: PluginContext,
  hookName: string,
): PluginContext => {
  if (!service.initialized) return pluginContext

  service.send(viteAdaptorModel.events.HOOK_START(hookName))

  const proxy = new Proxy(pluginContext, {
    get(target, key, receiver) {
      switch (key) {
        case 'emitFile':
          if (service.initialized) {
            return sendEmitFile
          }

        // eslint-disable-next-line no-fallthrough
        default:
          return Reflect.get(target, key, receiver)
      }
    },
  })

  return proxy
}
