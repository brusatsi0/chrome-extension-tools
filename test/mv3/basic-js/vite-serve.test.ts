import { fileWriteComplete } from '$src/viteAdaptor'
import fs from 'fs-extra'
import path from 'path'
import { createServer, ViteDevServer } from 'vite'

const outDir = path.join(__dirname, 'dist-serve')

let devServer: ViteDevServer
beforeAll(async () => {
  await fs.remove(outDir)

  devServer = await createServer({
    configFile: path.join(__dirname, 'vite.config.ts'),
    envFile: false,
    build: { outDir },
  })
})

afterAll(async () => {
  await devServer.close()
})

test('writes entry points to disk', async () => {
  expect(fs.existsSync(outDir)).toBe(false)

  await Promise.all([devServer.listen(), fileWriteComplete()])

  const { port } = devServer.config.server

  expect(fs.existsSync(outDir)).toBe(true)

  const manifest = 'manifest.json'
  const popup = 'popup.html'
  const content = 'content.js'
  const worker = 'service_worker.esm-wrapper.js'

  const manifestPath = path.join(outDir, manifest)
  const manifestSource = await fs.readJson(manifestPath)

  expect(manifestSource).toMatchObject({
    action: {
      default_popup: popup,
    },
    background: {
      service_worker: worker,
    },
    content_scripts: [
      {
        js: [content],
        matches: ['https://a.com/*', 'http://b.com/*'],
      },
    ],
  })

  const contentPath = path.join(outDir, content)
  const contentSource = await fs.readFile(contentPath, 'utf8')
  expect(contentSource).toMatch("console.log('content script')")

  const popupPath = path.join(outDir, popup)
  const popupSource = await fs.readFile(popupPath, 'utf8')
  expect(popupSource).toMatch(
    `<script src="http://localhost:${port}/popup.jsx" type="module">`,
  )

  const workerPath = path.join(outDir, worker)
  const workerSource = await fs.readFile(workerPath, 'utf8')
  expect(workerSource).toMatch(
    `import "http://localhost:${port}/service_worker.js"`,
  )
}, 60000)
