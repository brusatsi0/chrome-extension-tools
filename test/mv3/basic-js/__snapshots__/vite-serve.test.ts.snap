// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writes entry points to disk 1`] = `
"(function () {
  'use strict';

  (function() {
    console.log(\\"content script\\");
  })();

})();
"
`;

exports[`writes entry points to disk 2`] = `
"console.log(\\"service_worker.js\\");
"
`;
